zerver/lib/soft_deactivation.py

```diff
@@ -374,3 +375,31 @@ def get_soft_deactivated_users_for_catch_up(filter_kwargs: Any) -> List[UserProf
         **filter_kwargs,
     )
     return users_to_catch_up
+
+
+def soft_reactivate_if_personal_notification(
+    user_profile: UserProfile, unique_triggers: Set[str], mentioned_user_group_name: Optional[str]
+) -> None:
+    """When we're about to send an email/push notification to a
+    long_term_idle user, it's very likely that the user will try to
+    return to Zulip. As a result, it makes sense to optimistically
+    soft-reactivate that user, to give them a good return experience.
+
+    It's important that we do nothing for wildcard or group mentions,
+    because soft-reactivating an entire realm would be very expensive
+    (and we can't easily check the group's size). The caller is
+    responsible for passing a mentioned_user_group_name that is None
+    for messages that contain both a personal mention and a group
+    mention.
+    """
+    if not user_profile.long_term_idle:
+        return
+
+    private_message = "private_message" in unique_triggers
+    personal_mention = "mentioned" in unique_triggers and mentioned_user_group_name is None
+    if private_message or personal_mention:
```

Comment:
It's possible this would be better written as `if not private_mesage and not personal_mention: return None`. That would cause the test coverage system to push us to have tests for all the code paths.